File: src/main/java/com/regionrental/managers/EzChestShopManager.java
package com.regionrental.managers;

import com.regionrental.RegionRental;
import com.sk89q.worldedit.bukkit.BukkitAdapter;
import com.sk89q.worldedit.math.BlockVector3;
import com.sk89q.worldguard.protection.regions.ProtectedRegion;
import org.bukkit.Bukkit;
import org.bukkit.Location;
import org.bukkit.Material;
import org.bukkit.World;
import org.bukkit.block.Block;
import org.bukkit.plugin.Plugin;

import java.lang.reflect.Method;
import java.util.ArrayList;
import java.util.List;
import java.util.logging.Level;

/**
 * Manager for EzChestShop integration
 * Handles automatic shop removal from regions when rentals expire
 * Uses reflection to access EzChestShop's internal API (no compile-time dependency)
 */
public class EzChestShopManager {

    private final RegionRental plugin;
    private Plugin ezChestShopPlugin;
    private boolean ezChestShopEnabled;

    // Cached reflection objects for performance
    private Class<?> shopContainerClass;
    private Object shopContainerInstance;
    private Method isShopMethod;
    private Method deleteShopMethod;
    private Method hideForAllMethod;
    private boolean apiInitialized = false;

    public EzChestShopManager(RegionRental plugin) {
        this.plugin = plugin;
        setupEzChestShop();
    }

    /**
     * Initialize EzChestShop integration and reflection API
     */
    private void setupEzChestShop() {
        Plugin ecsPlugin = plugin.getServer().getPluginManager().getPlugin("EzChestShop");

        if (ecsPlugin == null) {
            plugin.getLogger().info("EzChestShop not found - shop removal disabled");
            this.ezChestShopEnabled = false;
            this.apiInitialized = false;
            return;
        }

        this.ezChestShopPlugin = ecsPlugin;
        this.ezChestShopEnabled = plugin.getConfigManager().isEzChestShopEnabled();

        if (!this.ezChestShopEnabled) {
            plugin.getLogger().info("EzChestShop detected but integration disabled in config");
            this.apiInitialized = false;
            return;
        }

        // Initialize reflection API
        this.apiInitialized = initializeReflectionAPI();

        if (this.apiInitialized) {
            plugin.getLogger().info("EzChestShop integration enabled (version: " + ecsPlugin.getDescription().getVersion() + ")");
        } else {
            plugin.getLogger().warning("EzChestShop found but API initialization failed - shop removal will be disabled");
            this.ezChestShopEnabled = false;
        }
    }

    /**
     * Initialize reflection objects for accessing EzChestShop's internal API
     *
     * @return true if initialization succeeded, false otherwise
     */
    private boolean initializeReflectionAPI() {
        try {
            // Load ShopContainer class
            shopContainerClass = Class.forName("me.deadlight.ezchestshop.data.ShopContainer");

            // Get ShopContainer instance from EzChestShop plugin
            Method getShopContainerMethod = ezChestShopPlugin.getClass().getMethod("getShopContainer");
            shopContainerInstance = getShopContainerMethod.invoke(ezChestShopPlugin);

            if (shopContainerInstance == null) {
                plugin.getLogger().warning("Failed to get ShopContainer instance from EzChestShop");
                return false;
            }

            // Cache the methods we'll need
            isShopMethod = shopContainerClass.getMethod("isShop", Location.class);
            deleteShopMethod = shopContainerClass.getMethod("deleteShop", Location.class);

            // Load ShopHologram class for hologram cleanup
            Class<?> shopHologramClass = Class.forName("me.deadlight.ezchestshop.utils.holograms.ShopHologram");
            hideForAllMethod = shopHologramClass.getMethod("hideForAll", Location.class);

            plugin.getLogger().fine("EzChestShop API reflection initialized successfully");
            return true;

        } catch (ClassNotFoundException e) {
            plugin.getLogger().warning("EzChestShop ShopContainer class not found - incompatible version?");
            return false;
        } catch (NoSuchMethodException e) {
            plugin.getLogger().warning("EzChestShop API methods not found - incompatible version: " + e.getMessage());
            return false;
        } catch (Exception e) {
            plugin.getLogger().log(Level.WARNING, "Failed to initialize EzChestShop API reflection", e);
            return false;
        }
    }

    /**
     * Check if EzChestShop is available, enabled, and API is initialized
     */
    public boolean isEnabled() {
        return ezChestShopEnabled && ezChestShopPlugin != null && ezChestShopPlugin.isEnabled() && apiInitialized;
    }

    /**
     * Remove all chest shops from a WorldGuard region
     * This is called during rental expiration after items have been stored
     *
     * @param regionName The name of the WorldGuard region
     * @return Number of shops removed, or -1 if operation failed
     */
    public int removeShopsInRegion(String regionName) {
        if (!isEnabled()) {
            plugin.getLogger().fine("EzChestShop integration not enabled, skipping shop removal");
            return 0;
        }

        // Get the WorldGuard region
        ProtectedRegion region = plugin.getWorldGuardManager().getRegion(regionName);
        if (region == null) {
            plugin.getLogger().warning("Cannot remove shops from region '" + regionName + "' - region not found");
            return -1;
        }

        // Find the world containing this region
        World world = findWorldForRegion(regionName);
        if (world == null) {
            plugin.getLogger().warning("Cannot remove shops from region '" + regionName + "' - world not found");
            return -1;
        }

        // Get all chest locations in the region
        List<Location> chestLocations = getChestLocationsInRegion(region, world);

        if (chestLocations.isEmpty()) {
            plugin.getLogger().fine("No chests found in region '" + regionName + "'");
            return 0;
        }

        // Remove shops from each chest location
        int shopsRemoved = 0;
        for (Location location : chestLocations) {
            if (removeShopAtLocation(location)) {
                shopsRemoved++;
            }
        }

        if (shopsRemoved > 0) {
            plugin.getLogger().info("Removed " + shopsRemoved + " chest shop(s) from region '" + regionName + "'");
        }

        return shopsRemoved;
    }

    /**
     * Find the world that contains the specified region
     */
    private World findWorldForRegion(String regionName) {
        for (World world : Bukkit.getWorlds()) {
            if (plugin.getWorldGuardManager().regionExists(regionName, world)) {
                return world;
            }
        }
        return null;
    }

    /**
     * Get all chest block locations within a WorldGuard region
     */
    private List<Location> getChestLocationsInRegion(ProtectedRegion region, World world) {
        List<Location> chestLocations = new ArrayList<>();

        // Get region bounds
        BlockVector3 min = region.getMinimumPoint();
        BlockVector3 max = region.getMaximumPoint();

        // Scan all blocks in the region
        for (int x = min.x(); x <= max.x(); x++) {
            for (int y = min.y(); y <= max.y(); y++) {
                for (int z = min.z(); z <= max.z(); z++) {
                    BlockVector3 blockVector = BlockVector3.at(x, y, z);

                    // Check if the block is actually in the region (handles irregular shapes)
                    if (!region.contains(blockVector)) {
                        continue;
                    }

                    Block block = world.getBlockAt(x, y, z);

                    // Check if block is a chest-type container that could be a shop
                    if (isChestType(block.getType())) {
                        chestLocations.add(block.getLocation());
                    }
                }
            }
        }

        return chestLocations;
    }

    /**
     * Check if a material is a chest-type block that can be used for shops
     */
    private boolean isChestType(Material material) {
        return material == Material.CHEST ||
               material == Material.TRAPPED_CHEST ||
               material == Material.BARREL;
    }

    /**
     * Remove a shop at a specific location using EzChestShop's internal API via reflection
     *
     * @param location The location of the chest
     * @return true if a shop was found and removed, false otherwise
     */
    private boolean removeShopAtLocation(Location location) {
        if (!apiInitialized) {
            return false;
        }

        try {
            // Check if this location has a shop
            Boolean isShop = (Boolean) isShopMethod.invoke(shopContainerInstance, location);

            if (isShop == null || !isShop) {
                // No shop at this location
                return false;
            }

            // STEP 1: Hide holograms for all players BEFORE deletion
            hideForAllMethod.invoke(null, location); // static method, pass null for instance

            // STEP 2: Delete the shop from database and memory
            deleteShopMethod.invoke(shopContainerInstance, location);

            // STEP 3: Hide holograms again as safety measure (handles async packet timing)
            hideForAllMethod.invoke(null, location);

            plugin.getLogger().fine("Successfully removed EzChestShop and holograms at " + formatLocation(location));
            return true;

        } catch (Exception e) {
            plugin.getLogger().log(Level.WARNING,
                    "Failed to remove shop at " + formatLocation(location) + " via reflection", e);
            return false;
        }
    }

    /**
     * Format a location for logging
     */
    private String formatLocation(Location loc) {
        return String.format("(%d, %d, %d) in %s",
                loc.getBlockX(), loc.getBlockY(), loc.getBlockZ(),
                loc.getWorld().getName());
    }

    /**
     * Reload the EzChestShop integration (called when config is reloaded)
     * Reinitializes reflection API in case plugin was updated
     */
    public void reload() {
        // Reset state
        this.apiInitialized = false;
        this.shopContainerClass = null;
        this.shopContainerInstance = null;
        this.isShopMethod = null;
        this.deleteShopMethod = null;
        this.hideForAllMethod = null;

        // Reinitialize
        setupEzChestShop();
    }
}

File: src/main/kotlin/com/zonerental/managers/EzChestShopManager.kt
package com.zonerental.managers

import com.zonerental.ZoneRental
import org.bukkit.Bukkit
import org.bukkit.Location
import org.bukkit.Material
import org.bukkit.World
import org.bukkit.plugin.Plugin
import java.lang.reflect.Method
import java.util.UUID
import java.util.logging.Level

/**
 * Manager for EzChestShop integration.
 * Uses reflection to access EzChestShop API and removes shops before region regeneration.
 */
class EzChestShopManager(private val plugin: ZoneRental) {

    private var ezChestShopEnabled = false

    // Reflection cache for shop detection and removal
    private var shopContainerClass: Class<*>? = null
    private var getShopMethod: Method? = null
    private var isShopMethod: Method? = null
    private var deleteShopMethod: Method? = null

    companion object {
        // Container types that can be EzChestShop shops
        private val SHOP_CONTAINER_TYPES = setOf(
            Material.CHEST,
            Material.TRAPPED_CHEST,
            Material.BARREL
        )

        private val POSSIBLE_PLUGIN_NAMES = arrayOf(
            "EzChestShop", "EzChestShopReborn", "ecs", "ChestShop", "ezchestshop"
        )
    }

    init {
        initializeReflection()
    }

    private fun initializeReflection() {
        // Try multiple possible plugin names (including Reborn variant)
        val ezChestShop = POSSIBLE_PLUGIN_NAMES
            .firstNotNullOfOrNull { Bukkit.getPluginManager().getPlugin(it) }

        if (ezChestShop == null) {
            plugin.logger.info("EzChestShop not detected - shop removal integration disabled")
            plugin.logger.info("Tried names: ${POSSIBLE_PLUGIN_NAMES.joinToString(", ")}")
            return
        }

        plugin.logger.info("Found EzChestShop plugin as: ${ezChestShop.name}")

        // Check if plugin is enabled - if not, schedule delayed initialization
        if (!ezChestShop.isEnabled) {
            plugin.logger.info("EzChestShop found but not yet enabled - will retry in 1 second...")
            Bukkit.getScheduler().runTaskLater(plugin, Runnable {
                if (ezChestShop.isEnabled) {
                    plugin.logger.info("EzChestShop now enabled - initializing integration...")
                    continueInitialization(ezChestShop)
                } else {
                    plugin.logger.warning("EzChestShop still not enabled after delay - integration disabled")
                }
            }, 20L) // 1 second delay
            return
        }

        // Plugin is already enabled - continue immediately
        continueInitialization(ezChestShop)
    }

    private fun continueInitialization(ezChestShop: Plugin) {
        plugin.logger.info("EzChestShop detected (v${ezChestShop.description.version}) - initializing integration...")

        try {
            shopContainerClass = Class.forName("me.deadlight.ezchestshop.data.ShopContainer")

            // Try different method names for compatibility
            getShopMethod = runCatching {
                shopContainerClass?.getMethod("getShop", Location::class.java)
            }.onFailure {
                plugin.logger.fine("getShop(Location) not found")
            }.getOrNull()

            isShopMethod = runCatching {
                shopContainerClass?.getMethod("isShop", Location::class.java)
            }.onFailure {
                plugin.logger.fine("isShop(Location) not found")
            }.getOrNull()

            // Try to find deleteShop method for shop removal
            deleteShopMethod = runCatching {
                shopContainerClass?.getMethod("deleteShop", Location::class.java)
            }.getOrElse {
                runCatching {
                    shopContainerClass?.getMethod("removeShop", Location::class.java)
                }.onFailure {
                    plugin.logger.warning("No deleteShop/removeShop method found - shop removal will not work")
                }.getOrNull()
            }

            if (getShopMethod != null || isShopMethod != null) {
                ezChestShopEnabled = true
                plugin.logger.info("EzChestShop integration enabled successfully")
                if (deleteShopMethod != null) {
                    plugin.logger.info("Shop removal API method available")
                }
            } else {
                plugin.logger.warning("EzChestShop integration failed - API methods not found")
            }

        } catch (e: ClassNotFoundException) {
            plugin.logger.warning("EzChestShop ShopContainer class not found: ${e.message}")
        } catch (e: Exception) {
            plugin.logger.log(Level.WARNING, "Failed to initialize EzChestShop integration", e)
        }
    }

    val isEnabled: Boolean
        get() = ezChestShopEnabled

    fun hasShopAt(location: Location): Boolean {
        if (!ezChestShopEnabled) return false

        return runCatching {
            isShopMethod?.let { method ->
                val result = method.invoke(null, location)
                if (result is Boolean) return result
            }
            getShopMethod?.let { method ->
                val shop = method.invoke(null, location)
                return shop != null
            }
            false
        }.onFailure { e ->
            if (plugin.configManager.isDebugMode) {
                plugin.logger.fine("Error checking shop at ${formatLocation(location)}: ${e.message}")
            }
        }.getOrDefault(false)
    }

    fun removeShopAt(location: Location): Boolean {
        if (!ezChestShopEnabled) return false

        val debug = plugin.configManager.isDebugMode

        if (!hasShopAt(location)) {
            if (debug) {
                plugin.logger.info("[Debug] No shop found at ${formatLocation(location)}")
            }
            return false
        }

        if (debug) {
            plugin.logger.info("[Debug] Removing shop at ${formatLocation(location)}")
        }

        // Call ShopContainer.deleteShop(location) via reflection
        // This handles: database removal, hologram cleanup, persistent data, memory cleanup
        val deleteMethod = deleteShopMethod
        if (deleteMethod != null) {
            return runCatching {
                deleteMethod.invoke(null, location)

                if (debug) {
                    plugin.logger.info("[Debug] Called deleteShop API at ${formatLocation(location)}")
                }

                // Verify shop was removed
                val removed = !hasShopAt(location)

                if (removed && debug) {
                    plugin.logger.info("[Debug] Shop successfully removed at ${formatLocation(location)}")
                } else if (!removed) {
                    plugin.logger.warning("Shop removal API call failed at ${formatLocation(location)}")
                }

                removed
            }.onFailure { e ->
                plugin.logger.warning("Failed to call deleteShop API: ${e.message}")
            }.getOrDefault(false)
        } else {
            plugin.logger.warning("deleteShop method not available - cannot remove shop at ${formatLocation(location)}")
            return false
        }
    }

    /**
     * Remove all chest shops from a WorldGuard region (world-aware).
     *
     * @param regionName The name of the WorldGuard region
     * @param world The world containing the region
     * @return Number of shops removed
     */
    fun removeShopsInRegion(regionName: String, world: World): Int {
        if (!ezChestShopEnabled) return 0

        if (!plugin.configManager.isEzChestShopRemovalEnabled) {
            return 0
        }

        // Get the WorldGuard region from the specified world
        val region = plugin.worldGuardManager.getRegion(regionName, world)
        if (region == null) {
            plugin.logger.warning("Cannot remove shops from region '$regionName' in world '${world.name}' - region not found")
            return 0
        }

        return removeShopsInRegionInternal(region, world, regionName)
    }

    /**
     * Remove all chest shops from a WorldGuard region (searches all worlds).
     * This method maintains the original signature for backward compatibility.
     *
     * @param regionName The name of the WorldGuard region
     * @return Number of shops removed
     */
    @Deprecated("Use removeShopsInRegion(String, World) instead")
    fun removeShopsInRegion(regionName: String): Int {
        if (!ezChestShopEnabled) return 0

        if (!plugin.configManager.isEzChestShopRemovalEnabled) {
            return 0
        }

        // Find the world containing this region
        val world = findWorldForRegion(regionName)
        if (world == null) {
            plugin.logger.warning("Cannot remove shops from region '$regionName' - world not found")
            return 0
        }

        // Get the WorldGuard region
        val region = plugin.worldGuardManager.getRegion(regionName, world)
        if (region == null) {
            plugin.logger.warning("Cannot remove shops from region '$regionName' - region not found")
            return 0
        }

        return removeShopsInRegionInternal(region, world, regionName)
    }

    /**
     * Internal method to remove shops from a region.
     */
    private fun removeShopsInRegionInternal(
        region: com.sk89q.worldguard.protection.regions.ProtectedRegion,
        world: World,
        regionId: String
    ): Int {
        val debug = plugin.configManager.isDebugMode
        val min = region.minimumPoint
        val max = region.maximumPoint

        if (debug) {
            plugin.logger.info("[Debug] Scanning region '$regionId' for shops")
        }

        // Find all shops
        val shopLocations = mutableListOf<Location>()
        for (x in min.x()..max.x()) {
            for (y in min.y()..max.y()) {
                for (z in min.z()..max.z()) {
                    val loc = Location(world, x.toDouble(), y.toDouble(), z.toDouble())
                    val block = loc.block
                    if (block.type in SHOP_CONTAINER_TYPES && hasShopAt(loc)) {
                        shopLocations.add(loc)
                    }
                }
            }
        }

        if (shopLocations.isEmpty()) {
            if (debug) {
                plugin.logger.info("[Debug] No shops found in region '$regionId'")
            }
            return 0
        }

        plugin.logger.info("Found ${shopLocations.size} EzChestShop(s) in region '$regionId'")

        // Remove all shops
        var removedCount = 0
        for (shopLoc in shopLocations) {
            if (removeShopAt(shopLoc)) {
                removedCount++
            }
        }

        plugin.logger.info("Removed $removedCount/${shopLocations.size} shop(s) from '$regionId'")
        return removedCount
    }

    /**
     * Find the world that contains the specified region.
     */
    private fun findWorldForRegion(regionName: String): World? {
        return Bukkit.getWorlds().firstOrNull { world ->
            plugin.worldGuardManager.regionExists(regionName, world)
        }
    }

    /**
     * Notify a player that shops were removed from their rental.
     */
    fun notifyPlayer(playerUUID: UUID, regionId: String, shopCount: Int) {
        if (!plugin.configManager.isEzChestShopNotifyEnabled) return

        Bukkit.getPlayer(playerUUID)?.let { player ->
            if (player.isOnline) {
                val message = plugin.configManager.ezChestShopRemovalMessage
                    .replace("{region}", regionId)
                    .replace("{count}", shopCount.toString())
                    .replace("&", "§")
                player.sendMessage(message)
            }
        }
    }

    private fun formatLocation(loc: Location): String {
        return "${loc.world.name} @ ${loc.blockX},${loc.blockY},${loc.blockZ}"
    }

    fun reload() {
        ezChestShopEnabled = false
        shopContainerClass = null
        getShopMethod = null
        isShopMethod = null
        deleteShopMethod = null
        initializeReflection()
    }
}
